### ISSUE TITLE
flox activate recovers dangling symlinks

### ISSUE THREAD

**Issue by ghudgins** [flox/flox#1394]
Current summary: flox should detect when environments have a dangling symlink and rebuild

### Reproduction case 
```
~/tmp
% nix run 'github:flox/flox/aaaaec3' -- init
✨ Created environment 'tmp' (aarch64-darwin)

Next:
  $ flox search <package>    <- Search for a package
  $ flox install <package>   <- Install a package into an environment
  $ flox activate            <- Enter the environment


~/tmp
% nix run 'github:flox/flox/aaaaec3' -- activate
✅ You are now using the environment 'tmp'.
To stop using this environment, type 'exit'

flox [tmp] 
(2) ~/tmp
% exit

~/tmp
% nix run 'github:flox/flox/v1.0.4' -- activate           
✅ You are now using the environment in current directory.
To stop using this environment, type 'exit'

/nix/store/96jv5cxlfmhl63291sxm5hjw5m50lsq2-flox.zdotdir/.zshrc:source:29: not a directory: /Users/dcarley/tmp/.flox/run/aarch64-darwin.tmp/activate/zsh

(2) ~/tmp
127 %
```

### Original bug 

STR:

* be on main (1.0.4 has slightly different behavior)
* `flox activate -r ghudgins/default`

Result: ❌ ERROR: No such file or directory (os error 2)

Expected: it works

Suspected root cause: dangling symlink in my cache directory

**Comment by ghudgins** [comment]
1.0.4 does this and activates:
<img width="582" alt="image" src="https://github.com/flox/flox/assets/5465514/3d1246f4-bdf4-43c4-8124-1fd9a6fdf496">

main generates an error:
<img width="592" alt="image" src="https://github.com/flox/flox/assets/5465514/de182c5c-f949-4af2-8280-4c797ec1686b">

**Comment by limeytexan** [comment]
I will look into this later today.

**Comment by limeytexan** [comment]
I'm not able to reproduce the problem (see below). I'm wondering if #1400 may have addressed the problem you were seeing?

```
brantley@Michaels-MacBook-Pro flox2 % result/bin/flox --version
1.0.3-70-ge3bbc8a
brantley@Michaels-MacBook-Pro flox2 % result/bin/flox activate -r ghudgins/default                   
✅ You are now using the environment 'ghudgins/default' (remote).
To stop using this environment, type 'exit'

Restored session: Fri  3 May 2024 17:32:14 CDT
flox [ghudgins/default (remote)] brantley@Michaels-MacBook-Pro flox2 % 

Saving session...
...saving history...truncating history files...
...completed.
brantley@Michaels-MacBook-Pro flox2 % SHELL=bash result/bin/flox activate -r ghudgins/default
✅ You are now using the environment 'ghudgins/default' (remote).
To stop using this environment, type 'exit'


The default interactive shell is now zsh.
To update your account to use zsh, please run `chsh -s /bin/zsh`.
For more details, please visit https://support.apple.com/kb/HT208050.
flox [ghudgins/default (remote)] bash-3.2$ exit
brantley@Michaels-MacBook-Pro flox2 % 
```

I've asked Graham for `-vvv` logs to see if I can see what's going on.

**Comment by limeytexan** [comment]
Working with Graham I was able to reproduce, but the reason for my failure is just a dangling symbolic link on account of the fact that I just copied Graham's cache files to my machine.
```
brantley@Michaels-MacBook-Pro flox2 % result/bin/flox -v activate -r ghudgins/default                                  
2024-05-04T13:28:44.034694Z  INFO flox::commands::activate: EXTRA DEBUG activate_path: /Users/brantley/.cache/flox/remote/ghudgins/default/.flox/run/activate    
✅ You are now using the environment 'ghudgins/default' (remote).
To stop using this environment, type 'exit'

❌ ERROR: No such file or directory (os error 2)
brantley@Michaels-MacBook-Pro flox2 % tracelinks /Users/brantley/.cache/flox/remote/ghudgins/default/.flox/run/activate
/Users/brantley/.cache/flox/remote/ghudgins/default/.flox/run -> /nix/store/91vg4gb40ahlvrdzaa3qq3zfz6zrccsm-environment
/nix/store/91vg4gb40ahlvrdzaa3qq3zfz6zrccsm-environment: No such file or directory
brantley@Michaels-MacBook-Pro flox2 % 
```
I've not yet been able to verify with Graham that this is the reason for his failure. (Although given that he's just returned from being away it's a little more likely than usual that the environment package did get garbage collected out from under him.)

If I instead turn the logic upside-down and instead use `! activate_path.is_executable()` to launch into the old logic then it fails as follows:
```
brantley@Michaels-MacBook-Pro flox2 % /Users/brantley/src/flox2/cli/target/debug/flox -v activate -r ghudgins/default
2024-05-04T13:46:39.908810Z  INFO flox::commands::activate: EXTRA DEBUG activate_path: /Users/brantley/.cache/flox/remote/ghudgins/default/.flox/run/activate    
✅ You are now using the environment 'ghudgins/default' (remote).
To stop using this environment, type 'exit'

Restored session: Fri  3 May 2024 18:10:22 CDT
/nix/store/14f338gzwfm1n3735ihbh8is88gklx1c-zdotdir/.zshrc:source:34: no such file or directory: /Users/brantley/.cache/flox/remote/ghudgins/default/.flox/run/activate/zsh
brantley@Michaels-MacBook-Pro flox2 % 
```

So either way I think this particular flavour of the bug can be summarised as "flox should detect when environments garbage collected out from under cache links" or something like that. If Graham can similarly confirm he's seeing a dangling `run` link then I think we have a good handle on the issue.

**Comment by dcarley** [comment]
I've experienced what apears to be the same issue, which would affect people downgrading back to v1.0.4 (for whatever reason). Simplest reproduction case:
```sh
~/tmp
% nix run 'github:flox/flox/aaaaec3' -- init
✨ Created environment 'tmp' (aarch64-darwin)

Next:
  $ flox search <package>    <- Search for a package
  $ flox install <package>   <- Install a package into an environment
  $ flox activate            <- Enter the environment


~/tmp
% nix run 'github:flox/flox/aaaaec3' -- activate
✅ You are now using the environment 'tmp'.
To stop using this environment, type 'exit'

flox [tmp] 
(2) ~/tmp
% exit

~/tmp
% nix run 'github:flox/flox/v1.0.4' -- activate           
✅ You are now using the environment in current directory.
To stop using this environment, type 'exit'

/nix/store/96jv5cxlfmhl63291sxm5hjw5m50lsq2-flox.zdotdir/.zshrc:source:29: not a directory: /Users/dcarley/tmp/.flox/run/aarch64-darwin.tmp/activate/zsh

(2) ~/tmp
127 %
```

**Comment by mkenigs** [comment]
It doesn't mean we can't recover dangling symlinks, but just noting that we just discovered a possible way to end up in this situation is by running `brew upgrade --cask`, which performs an uninstall and then an install. That completely wipes `/nix`.

**Comment by stahnma** [comment]
Another reproducer is to remove whatever your `$HOME/.cache/flox/run/*` points to and then try to activate that environment. This happens during a `brew upgrade flox` and it's a problem. This is a managed/remote env. 

remove whatever this points to ` .cache/flox/run/stahnma/default.5d13975f` (or whatever your env is) 

Then try to activate and your. get a "no such file or directory" type error.

**Comment by ysndr** [comment]
Fix merged in #1554
Thanks @nicklewis 
Is there anything else blocking this issue?

**Comment by ghudgins** [comment]
I don't think so, lets resolve and reopen if necessary.
